Add documentation for UI development (#65)
smartdevicelink/sdl_core_guides: docs/Integrating Your HMI/Developing the HMI UI/index.md

![App List](./assets/AppList1.png)

If an app is disconnected or unregistered, Core will send an `UpdateAppList` request to the HMI with that application omitted from the app list array.

The HMI should make sure its app list is always up to date, and only show applications that were included in the most recent `UpdateAppList` request.

## Activating an Application

### User Selection

When the user selects an application from the app list, a request should be made to bring this app to the foreground (this is called "activating" the application). The first step required by the HMI when an application is selected is to send a `SDL.ActivateApp` request to Core.

When Core responds with a successful `SDL.ActivateApp` response, the HMI can switch views from the app list to the app's default template.

!!! NOTE
The default template for an app should be used if the app has not requested to use a specific template via the `UI.Show.templateConfiguration` parameter or `UI.SetDisplayLayout` (deprecated).

The default template for media apps is `MEDIA`, and the default template for all other apps should be `NON-MEDIA`.

> You can check a given app is a media application using that app's `isMediaApplication` parameter, sent in the `BasicCommunication.UpdateAppList` request.

Reviewer: ```suggestion
You can check if a given app is a media application using that app's `isMediaApplication` parameter, sent in the `BasicCommunication.UpdateAppList` request.
```

!!!

### User Consent

If the `SDL.ActivateApp` response returns with the parameter `isPermissionsConsentNeeded = true`, the HMI should send a `SDL.GetListOfPermissions` request. This happens when the activating app requires permissions that the user must provide consent for. For example, if an app wants to access vehicle data, an SDL policy configuration might require that the user consent to allow the app to collect this information.

> After receiving the list of permissions for the app, the HMI should show the user the `PermissionItem` name and status for each requested permission. The user should have the ability to enable or disable each permission item. If any permission changes are made by the user, these updates should be communicated to core via `SDL.OnAppPermissionConsent` notification.

Reviewer: ```suggestion
After receiving the list of permissions for the app, the HMI should show the user the `PermissionItem` name and status for each requested permission. The user should have the ability to enable or disable each permission item. If any permission changes are made by the user, these updates should be communicated to Core via the `SDL.OnAppPermissionConsent` notification.
```

!!! NOTE
Permissions are managed by SDL Core's policy table. Refer to the [SDL Overview Policy Guide](https://smartdevicelink.com/en/guides/sdl-overview-guides/policies/overview/).
!!!

### Resumption

> If an app is disconnected from SDL Core and reconnects within a specified time limit, Core will try to resume the app into the same HMI state the app was in before resumption. The HMI should be prepared to handle a `BasicCommunication.ActivateApp` request from SDL Core, in which case the HMI should return the app into the requested state (or respond with an error if unable to). For example, if the requested HMI level is `FULL`, the HMI should activate the app and put that app's template into view.

Reviewer: >If an app is disconnected from SDL Core and reconnects within a specified time limit, Core will try to resume the app into the same HMI state the app was in before resumption. 

Should this instead end "...into the same HMI state the app was in before it was disconnected."?

!!! NOTE
`BasicCommunication.ActivateApp` is used differently than the previously described `SDL.ActivateApp`, but can be easily confused.

`SDL.ActivateApp` is a request originating from the HMI and should be sent when the user selects an app to activate.

`BasicCommunication.ActivateApp` is a request originating from SDL Core to move an app into a specific state. It is generally received by the HMI during app resumption.
!!!

## Displaying Information

> When an app wants to display information on the head unit, the HMI will receive a `UI.Show` request. The `UI.Show` request provides the HMI with the text, soft button information, and images an app has requested to display. The HMI should store the information in these requests for when an app activated and put into full. `UI.Show` requests are not always sent when an app is activated and in view.

Reviewer: ```suggestion
When an app wants to display information on the head unit, the HMI will receive a `UI.Show` request. The `UI.Show` request provides the HMI with the text, soft button information, and images an app has requested to display. The HMI should store the information in these requests for when an app is activated and put into full. `UI.Show` requests are not always sent when an app is activated and in view.
```

!!! NOTE

The HMI should merge the information in `UI.Show` requests with existing show information received for an app. 

If an HMI receives a request with text parameter `mainfield1` and a second request with text parameter `mainfield2`, the HMI should display both `mainfield1` and `mainfield2`.

If an app wants to clear a text field that it sent in a previous `UI.Show` request, SDL Core will send the HMI a request with that parameter's value set to an empty string ("").

!!!

## Implementing Soft Buttons

A `Softbutton` received from a `UI.Show` request should be displayed when the app is displaying a template. A template can have a max of 8 `Softbuttons`. These buttons can be of type `TEXT`, `IMAGE`, or `BOTH`.

> The HMI should keep an internal state of `SoftButtons` received by `UI.Show` requests, similar to how text fields and graphics are stored. Each `SoftButton` has a unique ID which must be saved by the HMI. These ID's are used in any messages sent to SDL Core when a user interacts with a `SoftButton`.

Reviewer: ```suggestion
The HMI should keep an internal state of `SoftButtons` received by `UI.Show` requests, similar to how text fields and graphics are stored. Each `SoftButton` has a unique ID which must be saved by the HMI. These IDs are used in any messages sent to SDL Core when a user interacts with a `SoftButton`.
```

> The actions expected of the HMI when the user selects a `SoftButton` are:

Reviewer: ```suggestion
The actions expected of the HMI when the user selects a `SoftButton` are:

```
You need an extra line break here for the bullet formatting to work.

>  - HMI Sends a notification `UI.OnButtonEvent` with buttonEventMode = `DOWN` when the user presses a button.

Reviewer: ```suggestion
 - HMI sends a notification `UI.OnButtonEvent` with buttonEventMode = `DOWN` when the user presses a button.
```

>  - HMI Sends a notification `UI.OnButtonEvent` with buttonEventMode = `UP` when the user releases a button.

Reviewer: ```suggestion
 - HMI sends a notification `UI.OnButtonEvent` with buttonEventMode = `UP` when the user releases a button.
```

>  - HMI Sends a notification `UI.OnButtonPress` with buttonPressMode = `SHORT` or `LONG`, depending on how long the user held the button in a down state.

Reviewer: ```suggestion
 - HMI sends a notification `UI.OnButtonPress` with buttonPressMode = `SHORT` or `LONG`, depending on how long the user holds the button in a down state.
```

!!! NOTE

> Not all HMIs support the ability to detect a button press duration, or differentiate between an up and down button event. In this case the HMI should make sure its ButtonCapabilities are accurately sent to core via the `ButtonCapabilities` parameter in `UI.GetCapabilities`. 

Reviewer: ```suggestion
Not all HMIs support the ability to detect a button press duration, or differentiate between an up and down button event. In this case the HMI should make sure its ButtonCapabilities are accurately sent to Core via the `ButtonCapabilities` parameter in `UI.GetCapabilities`. 
```

[More on HMI capabilities](#defining-the-ui-capabilities).
!!!

## Switching Templates

SDL Core can request the HMI to change an app's template using a `UI.Show` request, or the deprecated RPC `UI.SetDisplayLayout`.

In order to specify the template to be displayed, the `UI.Show` request uses the `templateConfiguration` parameter, which includes a string for the requested layout.

For `UI.SetDisplayLayout`, the `displayLayout` string parameter is used for the same purpose.

Using `UI.Show` is the preferred method because the request can be used to change the layout of the screen and the screen contents in a single request. This helps prevent lag and screen flashing when an app wants to change an app template.

An SDL app should only request to view templates that are supported in the HMI Capabilities. The HMI may return a failed response to Core in the event an unsupported template is requested. 

[More on HMI capabilities](#defining-the-ui-capabilities).

## Supported Template Views

> A reference list for all supported template views can be found [here](https://smartdevicelink.com/en/guides/sdl-overview-guides/user-interface/supported-templates/). This list shows screen shots of the 15 supported template views and how their text, graphic, and soft button components are arranged.

Reviewer: ```suggestion
A reference list for all supported template views can be found [here](https://smartdevicelink.com/en/guides/sdl-overview-guides/user-interface/supported-templates/). This list shows screenshots of the 15 supported template views and how their text, graphic, and soft button components are arranged.
```

> The defined strings for each template can be found in `PredefinedLayout` struct in the Mobile API RPC Specification. 

Reviewer: ```suggestion
The defined strings for each template can be found in the `PredefinedLayout` struct in the Mobile API RPC Specification. 
```

## Creating the App Menu

> Each application is able to maintain a list of menu commands through SDL. This in-app menu should be made accessible from an app's template view. Please note the example placement of the hamburger menu icon in the top right of the screen shot below.

Reviewer: ```suggestion
Each application is able to maintain a list of menu commands through SDL. This in-app menu should be made accessible from an app's template view. Please note the example placement of the hamburger menu icon in the top right of the screenshot below.
```

For `UI.SetDisplayLayout`, the `displayLayout` string parameter is used for the same purpose.

Using `UI.Show` is the preferred method because the request can be used to change the layout of the screen and the screen contents in a single request. This helps prevent lag and screen flashing when an app wants to change an app template.

An SDL app should only request to view templates that are supported in the HMI Capabilities. The HMI may return a failed response to Core in the event an unsupported template is requested. 

[More on HMI capabilities](#defining-the-ui-capabilities).

## Supported Template Views

A reference list for all supported template views can be found [here](https://smartdevicelink.com/en/guides/sdl-overview-guides/user-interface/supported-templates/). This list shows screenshots of the 15 supported template views and how their text, graphic, and soft button components are arranged.

The defined strings for each template can be found in the `PredefinedLayout` struct in the Mobile API RPC Specification. 

## Creating the App Menu

> Each application is able to maintain a list of menu commands through SDL. This in-app menu should be made accessible from an app's template view. Please note the example placement of the hamburger menu icon in the top right of the screenshot below.

Author: ![Screen Shot 2020-02-26 at 11 41 39 AM](https://user-images.githubusercontent.com/12038416/75366607-f9edd300-588c-11ea-850e-5ab8400fe485.png)

#### List Menu Example

![List Menu](./assets/list_menu.png)

#### Tiles Menu Example

![Tiles Menu](./assets/tiles_menu.png)

## Implementing Popups

There are several RPCs which are used to display a popup or an overlay to the user.

### UI.Alert

Alert is used to display a simple popup that can contain an image, text, and buttons.

![Alert](./assets/alert.png)

### UI.PerformInteraction

`PerformInteraction` is used to display a popup with contents which are displayed in a similar way to the app menu.

![Perform Interaction](./assets/perform_interaction.png)

### UI.Slider

`Slider` is used to display a popup that allows the user to enter a value via a slider input.

![Slider](./assets/slider.png)

### UI.ScrollableMessage

`ScrollableMessage` is used to display a popup which shows a long message to the user that requires scrolling.

![Scrollable Message](./assets/scrollable_message.png)

!!! NOTE

It is important that the HMI sends SDL Core a `UI.OnSystemContext` notification when displaying and closing a popup. A `systemContext` value of `ALERT` is used when `UI.Alert` is active, `HMI_OBSCURED` is used for all other popups.

!!!

## Navigating Through the IVI

> It is common for an SDL UI to be integrated into an existing OEM's UI. In order for SDL Core to work well with a head unit that has other embedded components, the HMI should make use of the `BasicCommunication.OnEventChanged` notification. This notification allows connected SDL applications to receive updates about their HMI status when a user interacts with other components like the embedded navigation or radio. For example, if an SDL media application is active and is playing audio, then the user switches the audio source to the embedded radio, the HMI should send SDL Core a `BasicCommunication.OnEventChanged` notification with `eventName = AUDIO_SOURCE` and `isActive = true`. This HMI notification will let the media application know that it no longer has control of the audio source.

Reviewer 2: ```suggestion
It is common for an SDL UI to be integrated into an existing OEM's UI. In order for SDL Core to work well with a head unit that has other embedded components, the HMI should make use of the `BasicCommunication.OnEventChanged` notification. This notification allows connected SDL applications to receive updates about their HMI status when a user interacts with other components like the embedded navigation or radio.

For example, if an SDL media application is active and is playing audio, then the user switches the audio source to the embedded radio, the HMI should send SDL Core a `BasicCommunication.OnEventChanged` notification with `eventName = AUDIO_SOURCE` and `isActive = true`. This HMI notification will let the media application know that it no longer has control of the audio source.
```
